### PR TITLE
GG-29382 [IGNITE-12891] Fix userAttributes sending for GridClient

### PR DESCRIPTION
**Pull request by isapego**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java [hunk 1]
```diff
@@ -29,34 +32,55 @@
...
     }
 
     /** {@inheritDoc} */
     @Override public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
         super.readExternal(in);
 
-        cred = in.readObject();
+        cred = (SecurityCredentials)in.readObject();
+
+        userAttrs = U.readMap(in);
```
Comment (antonovsergey93) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:83]: Is this change compatible? Can we get this request from the "old" node in a rolling upgrade scenario?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java [hunk 2]
```diff
@@ -29,34 +32,55 @@
...
      */
-    public void credentials(Object cred) {
+    public void credentials(SecurityCredentials cred) {
         this.cred = cred;
     }
 
+    /**
+     * @return User attributes.
+     */
+    public Map<String, String> userAttributes() {
```
Comment (antonovsergey93) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:57]: Please add `@Nullable` annotation to the return type.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java [hunk 3]
```diff
@@ -29,34 +32,55 @@
     private static final long serialVersionUID = 0L;
 
     /** Credentials. */
-    private Object cred;
+    private SecurityCredentials cred;
+
+    /** User attributes. */
+    private Map<String, String> userAttrs;
```
Comment (antonovsergey93) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:38]: Please add `@Nullable` annotation

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (antonovsergey93) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:2]: Please update copyright year in your IDE and all new classes.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        Ignite ignite = startGrids(2);
+
+        assertEquals(2, ignite.cluster().topologyVersion());
+
+        ignite.cluster().active(true);
+
+        try (GridClient client = GridClientFactory.start(getGridClientConfiguration())) {
+            assertTrue(client.connected());
+
+            assertEquals(userAttrs.get("key"), "val");
```
Comment (antonovsergey93) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:66]: Please look at JavaDoc. The first argument is expected, the second is actual.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+            // Close a coordinator to force the client to send a CLUSTER_CURRENT_STATE message to the other node
+            // in the state.state() statement.
+            ignite.close();
+
+            userAttrs = null;
+
+            state.active();
+
+            assertEquals(userAttrs.get("key"), "val");
```
Comment (antonovsergey93) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:78]: Please look at JavaDoc. The first argument is expected, the second is actual.

**Review by antonovsergey93 — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java (hunk 1 above)
Comment (isapego) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:83]: I don't know if it's compatible. I thought you have added tests to check that?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java (hunk 2 above)
Comment (isapego) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:57]: Done.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java (hunk 4 above)
Comment (isapego) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:2]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java (hunk 3 above)
Comment (isapego) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/client/message/GridClientAuthenticationRequest.java:38]: Done.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java (hunk 5 above)
Comment (isapego) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:66]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java (hunk 6 above)
Comment (isapego) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/client/AttributeSecurityCheckTest.java:78]: Fixed.

**Pull request closed without merging**